app/src/verifier.h

```diff
@@ -130,15 +132,243 @@ namespace scitt::verifier
+      {
+        throw cose::COSEDecodeError(
+          "cty must be 'application/vnd.cncf.notary.payload.v1+json' for "
+          "Notary claims");
+      }
+
+      if (notary_signing_scheme.value() == "notary.x509")
+      {
+        // notary_signing_time is not critical but is required iff notary.x509
+        if (!notary_signing_time.has_value())
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.signingTime in protected header");
+        }
+        if (notary_authentic_signing_time.has_value())
+        {
+          throw cose::COSEDecodeError(
+            "io.cncf.notary.authenticSigningTime not allowed in protected "
+            "header when io.cncf.notary.signingScheme is `notary.x509`");
+        }
+      }
+      else if (notary_signing_scheme.value() == "notary.x509.signingAuthority")
+      {
+        // notary_authentic_signing_time is critical and required iff
+        // notary.x509.signingAuthority
+        if (notary_signing_time.has_value())
+        {
+          throw cose::COSEDecodeError(
+            "Notary io.cncf.notary.signingTime not allowed in protected header "
+            "when io.cncf.notary.signingScheme is "
+            "`notary.x509.signingAuthority`");
+        }
+        if (!notary_authentic_signing_time.has_value())
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.authenticSigningTime in protected header");
+        }
+        if (
+          std::find(
+            crit.begin(),
+            crit.end(),
+            std::variant<int64_t, std::string>(std::string(
+              "io.cncf.notary.authenticSigningTime"))) == crit.end())
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.authenticSigningTime in crit parameters");
+        }
+      }
+      else
+      {
+        throw cose::COSEDecodeError(
+          "Notary io.cncf.notary.signingScheme must be `notary.x509` or "
+          "`notary.x509.signingAuthority`");
+      }
+
+      if (notary_expiry.has_value())
+      {
+        // notary_expiry is critial but not required.
+        if (
+          std::find(
+            crit.begin(),
+            crit.end(),
+            std::variant<int64_t, std::string>(
+              std::string("io.cncf.notary.expiry"))) == crit.end())
+        {
+          throw cose::COSEDecodeError(
+            "Missing io.cncf.notary.expiry in crit parameters");
+        }
+      }
+    }
+
+    PublicKey process_notary_profile(
+      const std::vector<uint8_t>& data,
+      cose::ProtectedHeader& phdr,
+      kv::Tx& tx,
+      const Configuration& configuration)
+    {
+      // Validate protected header
+      validate_notary_protected_header(phdr, configuration);
 
+      // Notary puts the x5chain in unprotected header,
+      // hence decode unprotected header.
+      cose::UnprotectedHeader unphdr;
+      try
+      {
+        unphdr = cose::decode_unprotected_header(data);
```

Comment:
We should decode the protected and unprotected headers at the same time IMO (it's fine to keep them as separate structs).
Have one `decode_headers(std::span<const uint8_t> cose_sign1)` function, which calls both `decode_protected_header(QCBORDecodeContext* ctx)` and `decode_unprotected_header(QCBORDecodeContext* ctx)`.